spec/section/source-vocabulary.md

```diff
@@ -456,3 +456,30 @@ specified by a string Literal. The file's absolute path is `/root/file.xml`.
   rml:iterator "$";
 .
 </pre>
+
+In some cases, it is useful to describe access to a source that will be programmatically
+provided to an RML processor at runtime.
+This could, for example, be a byte stream or an already deserialized JSON or XML node.
+
+For these cases, the `rml:ProvidedSource` can be used.
+It is a sub class of `rml:Source` and MUST have:
+
+* exactly one `rml:sourceIdentifier` property, whose value is a string literal, which
+uniquely identifies the source.
+
+| Property                | Domain                | Range      |
+| ------------------------| --------------------- | ---------- |
+| `rml:sourceIdentifier`  | `rml:ProvidedSource`  | `Literal`  |
+
+The following example illustrates a provided source with its unique identifier `myByteStream`:
```

Comment:
Don't we need a reference formulation? How does the processor know how to deal with that kind of ProvidedSource?
So an iterator + reference formulation? 

I guess the actual dealing is then defined in rml-io-registry once we have an iterator + reference formulation.

Reply:
Ah good point. The idea is that the reference formulation can be any reference formulation supported by the RML processor. It is up to the mapping author to choose the correct one that will match the provided source at runtime. So this addition has no impact on that. I agree that this is not very clear from the text and the example. I will clarify.

Reply:
Yeah any formulation is fine, but we need to clarify that somehow that you probably need one and if not provided, we fallback to the default of a Logical Source.

Reply:
I'm not sure I follow. What is the "default [reference formulation] of a Logical Source"?

What I mean is:
say you have processed some data in some runtime process. The idea is to be able to provide that to an RML processor during that runtime. The type of data that is provided should match one of the reference formulations supported by the RML processor. The type of reference formulation for this specific source is determined, as per usual, by the `rml:referenceFormulation` of the Logical Source. All we need is to know which data matches which provided source in the mapping. That's why we need a `rml:sourceIdentifier`. This identifier will also be provided to the RML processor together with the source data programmatically.

So say you process that data to some structure on which the engine can execute JSONPath queries. Then a mapping could look like:

```ttl
<> rml:logicalSource [
  rml:source [
    rml:sourceIdentifier "my-json-source" ;
  ] ;
  rml:iterator "$" ;
  rml:referenceFormulation rml:JSONPath ;
] ;
...
```

Reply:
Yes! Looks what I have mind!

> I'm not sure I follow. What is the "default [reference formulation] of a Logical Source"?

I was wrong, that's old RML. RML-Core now requires a reference formulation:

> exactly one rml:referenceFormulation property
https://kg-construct.github.io/rml-core/spec/docs/#logicalSource

Old RML had a default 'tabular' for this to be compatible with R2RML.